python/lsst/daf/butler/remote_butler/_remote_butler.py

```diff
@@ -216,19 +219,45 @@ def get(
         **kwargs: Any,
     ) -> Any:
         # Docstring inherited.
-        if not isinstance(datasetRefOrType, DatasetRef):
-            raise NotImplementedError("RemoteButler currently only supports get() of a DatasetRef")
-
-        dataset_id = datasetRefOrType.id
-        request = GetFileRequestModel(dataset_id=dataset_id)
-        response = self._post("get_file", request)
-        if response.status_code == 404:
-            raise LookupError(f"Dataset not found with ID ${dataset_id}")
+        if isinstance(datasetRefOrType, DatasetRef):
+            dataset_id = datasetRefOrType.id
+            response = self._get(f"get_file/{dataset_id}")
+            if response.status_code == 404:
+                raise LookupError(f"Dataset not found with ID {dataset_id}")
+        else:
+            request = GetFileByDataIdRequestModel(
+                dataset_type_name=self._normalize_dataset_type_name(datasetRefOrType),
+                collections=self._normalize_collections(collections),
+                data_id=self._simplify_dataId(dataId, kwargs),
+            )
+            response = self._post("get_file_by_data_id", request)
+            if response.status_code == 404:
+                raise LookupError(f"Dataset not found with DataId {dataId}")
+
         response.raise_for_status()
         model = self._parse_model(response, GetFileResponseModel)
 
+        # If the caller provided a DatasetRef or DatasetType, they may have
+        # overridden the storage class on it.  We need to respect this, if they
+        # haven't asked to re-override it.
+        explicitDatasetType = _extract_dataset_type(datasetRefOrType)
+        if explicitDatasetType is not None:
+            if storageClass is None:
+                storageClass = explicitDatasetType.storageClass
+
+        # If the caller provided a DatasetRef, they may have overridden the
+        # component on it.  We need to explicitly handle this because we did
+        # not send the DatasetType to the server in this case.
+        componentOverride = None
+        if isinstance(datasetRefOrType, DatasetRef):
```

Comment:
I realize that we talked about this not being a hotspot, but doing the instance check twice when we already know the answer a few lines higher up seems wasteful. Can't we calculate componentOverride in the block above?

Reply:
That is how I had it originally actually, but I re-organized it for clarity.  This version moves the declaration of variables closer to the code that uses them and separate the concerns.  Above the block is concerned with deciding which REST endpoint to call, down here we are focused on post-processing the result.  

I originally had both the storage class and component override thing up above, but I discovered that above there are only two cases (DatasetRef or not), but for postprocessing there are three cases (DatasetRef, inflated DatasetType, string DatasetType).

As I have been reading the existing Butler methods to try to understand how they work, one of the things that has making it more difficult is jumping up and down multi-hundred-line functions to figure out where a variable came from and whether it has been mutated since its original definition.  Unfortunately python lacks the `const` declarations that most other languages have.  And because Butler methods all take extremely long argument lists with variable argument types, it's harder to add helper functions to break them into chunks because all the helper methods would need to take like 6 parameters.

Instance checks are not expensive.  It would be better if the Butler API was designed in a way that didn't make them necessary, but unfortunately it's not.  In the absolute worst case it's walking a very short list (like 3 items) and doing a couple of method calls to `__isinstance__`.  In the typical case (which this is) it's basically just a pointer comparison.  It's not worth worrying about these kinds of microoptimizations except in tight inner loops.  This is not a tight inner loop, since this function does network I/O to go out to a service that then does network I/O to ask a database to do disk I/O.

Computers are pretty fast now, but developer time is more expensive than ever.  This kind of stuff is not worth worrying about.  Clarity is a lot more important, as is minimizing the amount of time spent arguing over things that don't matter so that there is more time available for writing code.